Lib/test/test_typing.py

```diff
@@ -3738,6 +3738,18 @@ def test_repr(self):
 
         self.assertEqual(repr(UserId), f'{__name__}.UserId')
 
+    def test_pickle(self):
```

Comment:
I was gonna suggest comparing equality of `s` and `loaded` but interestingly they're not actually treated equal:
```pycon
>>> g = NewType('G', int)
>>> g2 = pickle.loads(pickle.dumps(g))
>>> g == g2
False
```

It sorta makes sense but I wonder what @JelleZijlstra and @Fidget-Spinner think about it.

Reply:
Thats because `NewType` does not have `__eq__` method implemented. Should we implement it?

Reply:
Yup, I know, this isn't the problem. Consider this example:

```py3
# file: other.py
import typing

G = typing.NewType("G", int)
```

```py3
# file main.py
import typing
import other

G = typing.NewType("G", int)

def func(g: G) -> None:
    ...

if __name__ == "__main__":
    func(other.G(1))
```

Mypy raises:
```
main.py:10: error: Argument 1 to "func" has incompatible type "other.G"; expected "main.G"
Found 1 error in 1 file (checked 1 source file)
```

So clearly they're not treated as equivalent by static type checkers. PEP 484 doesn't specify what should happen with two NewType objects with the same name argument and wrapped type. However, Mypy at least is pretty strict about them, consider if we choose a different variable name for our NewType in `other.py`:

```
# other.py
import typing

G1 = typing.NewType("G", int)
```

This generates:
```
other.py:3: error: String argument 1 "G" to NewType(...) does not match variable name "G1"
Found 1 error in 1 file (checked 1 source file)
```

This is why I'd like to ask others for opinion since it's not at all clear whether multiple NewType objects with the same state should be treated equal. Static type checkers want:
- the origin module to be the same;
- the name of the variable to be the same as the name passed when creating the NewType object.

At the same time, pickling NewType objects (and most other type constructs) is kinda pointless from the static analysis angle so maybe runtime behavior shouldn't follow static type checking behavior here? Again, I'm not sure.

Reply:
@ambv Thanks for explanation. Let's wait for others opinion.

Reply:
I'd say we should keep backwards compatible behavior and delegate all the dunders to the callable.

Consider this code in 3.9:
```python
Foo = NewType('Foo', int)
Bar = NewType('Bar', int)

Foo == Bar #True! because int == int
```

Now not anymore.
So I propose ``__eq__`` (and any required dunder), be like this:
```python
def __eq__(self, other):
    return self.__supertype__ == other
```

Reply:
The problem with equality for the static type checker is harder to solve though.

Eg. `TypeVar('T') == TypeVar('T')` is False! Because even with the same name, they aren't the same to the static checker. And the runtime also reflects that, `TypeVar` equality is checked using object id.

Reply:
Your example `__eq__` code would be insufficient, it wouldn't make `Foo == Bar`.

But that doesn't matter. Looking at your code I realized something. The idea to keep backwards compatibility with pre-3.10 sounds good at first glance but it's problematic: because of hashability requirements we'd have to also make `hash(Foo) == hash(Bar)`. That would lead to a weird scenario where:
```pycon
>>> id = NewType('ID', int)
>>> age = NewType('Age', int)
>>> types = {id, age}
>>> len(types)
1
```

That is clearly wrong. So backwards compatibility isn't really an option.

In any case, we're only left with the question: should two NewTypes **with identical internal state** be considered equal.

Reply:
> Your example `__eq__` code would be insufficient, it wouldn't make `Foo == Bar`.

Woops, I meant to write ``return self.__supertype__ == other.__supertype__``. I should check my own code next time before commenting ;-).

> But that doesn't matter. Looking at your code I realized something. ...
> That is clearly wrong. So backwards compatibility isn't really an option.

Oh right. I also remembered that if things in Python are equal, then they must also (usually) produce the same hash. And this will become very messy if we allow that. Thanks for pointing that out.

> In any case, we're only left with the question: should two NewTypes **with identical internal state** be considered equal.

I suddenly realized I'd confused and misunderstood a lot of things about NewType. Sorry for the confusion. I forgot that old typing.NewType returned an identity function which couldn't be pickled, and its address would always be different So we shouldn't have any backwards incompatibility problems (I think). And equality comparisons would usually return `False` because of the different memory locations:

`NewType('', int) == NewType('', int) # False in 3.9`

For your question, I need to look at patterns of how people use ``NewType``. I honestly don't know what's the right approach right now :(. A bandaid would be keep the status quo -- it's been like this for 5 versions and users haven't complained about equality comparisons.

Reply:
One realistic use case where this could make a difference is in pickling functions that take NewType arguments. `cloudpickle` passes function objects around that way, I believe. For that use case, I don't think it actually matters: the function itself won't compare equal anyway after being unpickled.

Runtime type checkers looking at NewTypes would just access the `.__supertype__` and other attributes, and that would work OK either way. Making things equal that didn't previously compare equal could perhaps break some runtime type checkers.

Reply:
Yeah, existing runtime usage received a new function with each call to `NewType` although I doubt that anything would break if we made them **properly** comparable now (essentially by making `__eq__` do what `vars(o1) == vars(o2)` does because we want `__module__` to be part of the equality comparison too, as static type checkers do).

`TypeVar("T")` could be our guide in how to design `NewType` because it's got interesting properties:
- trying to pickle it, it looks up its module's globals for a `T` name and pickles *that* if it's the same as the given object;
- if you manually create two type vars with the same name, they are not considered equal;
- unpickling `TypeVar("T")` returns *the exact same object* that already lives as `module.T`;
- unpickling `TypeVar("T")` fails if there isn't a T in its module's globals.

I think it would be wise to duplicate this pickling behavior in `NewType`.

Reply:
+1 to copy TypeVar's behavior.

Reply:
Add `NewType.__reduce__` method to match logic of `TypeVar.__reduce__`

Reply:
If i remember correctly, you also need to copy some of TypeVar's `__init__`'s sys._getframe hacks to implement what Lukasz was talking about.

Reply:
It has been implemented. Please, see this line:

https://github.com/python/cpython/blob/83d1430ee5b8008631e7f2a75447e740eed065c1/Lib/typing.py#L2390

Reply:
Great thanks!